pinot-common/src/main/java/org/apache/pinot/common/concurrency/AdjustableSemaphore.java

```diff
@@ -20,32 +20,44 @@
 
 import com.google.common.base.Preconditions;
 import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
 
 
 /**
  * A semaphore that allows adjusting the number of permits in a non-blocking way.
  */
 public class AdjustableSemaphore extends Semaphore {
 
-  private int _totalPermits;
+  private final AtomicInteger _totalPermits;
 
   public AdjustableSemaphore(int permits) {
     super(permits);
-    _totalPermits = permits;
+    _totalPermits = new AtomicInteger(permits);
   }
 
   public AdjustableSemaphore(int permits, boolean fair) {
     super(permits, fair);
-    _totalPermits = permits;
+    _totalPermits = new AtomicInteger(permits);
   }
 
+  /**
+   * Sets the total number of permits to the given value without blocking.
+   */
   public void setPermits(int permits) {
     Preconditions.checkArgument(permits > 0, "Permits must be a positive integer");
-    if (permits < _totalPermits) {
-      reducePermits(_totalPermits - permits);
-    } else if (permits > _totalPermits) {
-      release(permits - _totalPermits);
+    if (permits < _totalPermits.get()) {
+      reducePermits(_totalPermits.get() - permits);
+    } else if (permits > _totalPermits.get()) {
+      release(permits - _totalPermits.get());
```

Comment:
This is confusing. If this method can be called concurrently, it doesn't seem we are protecting ourself agains races. If it cannot be called concurrently, it seems odd to use an atomic integer

Reply:
You're right, we can't actually have concurrent modifications. Only potential concurrent reads during modifications. I think we can simply use a volatile int instead (I guess even the volatile is not strictly necessary as we can tolerate stale values without many issues tbh).

Reply:
That is what I thought. Volatile should be enough to be correct. In practice, you could even remove the volatile as you suggest, but in theory, that could imply that the reader threads never read the new value. Also, IIRC, in x86, to read a volatile attribute that was not modified is exactly as to read a nonvolatile attribute. The only actual issue is if that attribute ends up creating false sharing with another, frequently written, not volatile attribute

Reply:
Interesting, thanks for that info!

> The only actual issue is if that attribute ends up creating false sharing with another, frequently written, not volatile attribute

Wouldn't the other attribute also need to be volatile?

Reply:
I was wrong. The issue is when a frequently written volatile is in the same cache as a non-volatile frequently read attribute. Each writes into the volatile will evict the cache line, so other threads reading the non-volatile will need to read the line again (https://medium.com/@ali.gelenler/cache-trashing-and-false-sharing-ce044d131fc0)